src/tools/particles/Controls.tsx

```diff
@@ -0,0 +1,124 @@
+import * as React from 'react'
+
+import { Config } from '~/tools/particles/interfaces'
+
+interface ConfigWithBit extends Config {
+  dirty: boolean
+}
+
+export const Controls = (props: {
+  initialConfig: Config
+  updateGlobalConfig: (c: Config) => void
+}) => {
+  const [config, setConfig] = React.useState<ConfigWithBit>({
+    ...props.initialConfig,
+    dirty: false,
+  })
+
+  const updateConfig = (key: string, value: any) => {
+    setConfig({ ...config, [key]: value, dirty: true })
+  }
+
+  const publishConfig = () => {
+    props.updateGlobalConfig(config)
+    setConfig({ ...config, dirty: false })
+  }
+
+  const parse = (val: string) => {
+    if (val === '') return ''
+    return parseFloat(val)
+  }
+
+  return (
+    <div style={{ margin: 'auto', width: 300 }}>
+      <div>
+        <label htmlFor="backgroundColor">Background Color</label>
+        <input
+          type="text"
+          name="backgroundColor"
+          value={config.backgroundColor}
+          onChange={(e) => {
+            updateConfig('backgroundColor', e.target.value)
+          }}
+        ></input>
+      </div>
+
+      <div>
+        <label htmlFor="emitterLifespan">Emitter lifespan</label>
+        <input
+          type="number"
+          name="emitterLifespan"
+          value={config.emitterLifespan}
+          onChange={(e) => {
+            updateConfig('emitterLifespan', parse(e.target.value))
+          }}
+        ></input>
+      </div>
+
+      <div>
+        <label htmlFor="orientation">Particle lifespan</label>
+        <input
+          type="number"
+          name="particleLifespan"
+          value={config.particleLifespan}
+          onChange={(e) => {
+            updateConfig('particleLifespan', parse(e.target.value))
+          }}
+        ></input>
+      </div>
+
+      <div>
+        <label htmlFor="orientation">Orientation</label>
+        <input
+          type="range"
+          name="orientation"
+          min="0"
+          max={Math.PI * 2}
+          step="0.1"
+          value={config.orientation}
+          onChange={(e) => {
+            updateConfig('orientation', parse(e.target.value))
+          }}
+        ></input>
+      </div>
+
+      <div>
+        <label htmlFor="arc">Arc</label>
+        <input
+          type="range"
+          name="arc"
+          min="0"
+          max={Math.PI * 2}
+          step="0.1"
+          value={config.arc}
+          onChange={(e) => {
+            updateConfig('arc', parse(e.target.value))
+          }}
+        ></input>
+      </div>
+
+      <button
+        type="submit"
```

Comment:
How would you feel about immediately applying the changes to the rendered effects? I think this would make the tool into something that gave you immediate feedback, which is really powerful.

Reply:
I was 50/50 on that so built it the other way, but let's trying it that way instead. Since it doesn't apply til the next emitter loop, I think it'll feel really good.